File: Cyaim.WebSocketServer/Clients/cyaim-websocket-client-java/src/main/java/com/cyaim/websocket/WebSocketClientFactory.java
package com.cyaim.websocket;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Factory for creating WebSocket client proxies based on server endpoints
 * 基于服务器端点创建 WebSocket 客户端代理的工厂
 */
public class WebSocketClientFactory {
    private final String serverBaseUrl;
    private final String channel;
    private final WebSocketClientOptions options;
    private List<WebSocketEndpointInfo> cachedEndpoints;
    private final OkHttpClient httpClient = new OkHttpClient();
    private final Gson gson = new Gson();

    public WebSocketClientFactory(String serverBaseUrl, String channel, WebSocketClientOptions options) {
        this.serverBaseUrl = serverBaseUrl.replaceAll("/$", "");
        this.channel = channel;
        this.options = options != null ? options : new WebSocketClientOptions();
    }

    /**
     * Get endpoints from server / 从服务器获取端点
     */
    public CompletableFuture<List<WebSocketEndpointInfo>> getEndpoints() {
        if (cachedEndpoints != null) {
            return CompletableFuture.completedFuture(cachedEndpoints);
        }

        return CompletableFuture.supplyAsync(() -> {
            try {
                String url = serverBaseUrl + "/ws_server/api/endpoints";
                Request request = new Request.Builder().url(url).build();
                
                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new IOException("Unexpected code: " + response);
                    }

                    String json = response.body().string();
                    Type type = new TypeToken<ApiResponse<List<WebSocketEndpointInfo>>>(){}.getType();
                    ApiResponse<List<WebSocketEndpointInfo>> apiResponse = gson.fromJson(json, type);

                    if (apiResponse.success && apiResponse.data != null) {
                        cachedEndpoints = apiResponse.data;
                        return cachedEndpoints;
                    }

                    throw new RuntimeException(apiResponse.error != null ? apiResponse.error : "Failed to fetch endpoints");
                }
            } catch (Exception e) {
                throw new RuntimeException("Error fetching WebSocket endpoints from server: " + e.getMessage(), e);
            }
        });
    }

    /**
     * Create a client / 创建客户端
     */
    public WebSocketClient createClient() {
        String wsUri = serverBaseUrl
            .replace("http://", "ws://")
            .replace("https://", "wss://");
        return new WebSocketClient(wsUri, channel);
    }

    private static class ApiResponse<T> {
        public boolean success;
        public T data;
        public String error;
    }
}

File: Cyaim.WebSocketServer/Clients/cyaim-websocket-client-dart/lib/src/websocket_client_factory.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'websocket_client.dart';
import 'websocket_client_options.dart';
import 'websocket_endpoint_info.dart';

/// Factory for creating WebSocket client proxies based on server endpoints
/// 基于服务器端点创建 WebSocket 客户端代理的工厂
class WebSocketClientFactory {
  final String serverBaseUrl;
  final String channel;
  final WebSocketClientOptions options;
  List<WebSocketEndpointInfo>? _cachedEndpoints;

  WebSocketClientFactory(this.serverBaseUrl, this.channel, [WebSocketClientOptions? options])
      : options = options ?? WebSocketClientOptions();

  /// Get endpoints from server / 从服务器获取端点
  Future<List<WebSocketEndpointInfo>> getEndpoints() async {
    if (_cachedEndpoints != null) {
      return _cachedEndpoints!;
    }

    try {
      final url = '${serverBaseUrl.replaceAll(RegExp(r'/$'), '')}/ws_server/api/endpoints';
      final response = await http.get(Uri.parse(url));

      if (response.statusCode != 200) {
        throw Exception('HTTP error! status: ${response.statusCode}');
      }

      final apiResponse = jsonDecode(response.body) as Map<String, dynamic>;
      
      if (apiResponse['success'] == true && apiResponse['data'] != null) {
        final data = apiResponse['data'] as List<dynamic>;
        _cachedEndpoints = data
            .map((e) => WebSocketEndpointInfo.fromJson(e as Map<String, dynamic>))
            .toList();
        return _cachedEndpoints!;
      }

      throw Exception(apiResponse['error'] ?? 'Failed to fetch endpoints');
    } catch (e) {
      throw Exception('Error fetching WebSocket endpoints from server: $e');
    }
  }

  /// Create a client / 创建客户端
  WebSocketClient createClient() {
    final wsUri = serverBaseUrl
        .replaceAll('http://', 'ws://')
        .replaceAll('https://', 'wss://');
    return WebSocketClient(wsUri, channel, options);
  }
}

